design/concurrent-backup-processing.md

```diff
@@ -0,0 +1,245 @@
+# Concurrent Backup Processing
+
+This enhancement will enable Velero to process multiple backups at the same time. This is largely a usability enhancement rather than a performance enhancement, since the overall backup throughput may not be significantly improved over the current implementation, since we are already processing individual backup items in parallel. It is a significant usability improvement, though, as with the current design, a user who submits a small backup may have to wait significantly longer than expected if the backup is submitted immediately after a large backup.
+
+## Background
+
+With the current implementation, only one backup may be `InProgress` at a time. A second backup created will not start processing until the first backup moves on to `WaitingForPluginOperations` or `Finalizing`. This is a usability concern, especially in clusters when multiple users are initiating backups. With this enhancement, we intend to allow multiple backups to be processed concurrently. This will allow backups to start processing immediately, even if a large backup was just submitted by another user. This enhancement will build on top of the prior parallel item processing feature by sharing the same item worker threads across the running backups. Those threads are already self-contained and designed to communicate safely with other goroutines via channels, and the threads themselves contain no internal references to particular backups.
+
+The primary challenge is to make sure that two overlapping backups are not processed concurrently. If that were to happen, we would risk data corruption, especially around the processing of pod hooks and volume backup. For the initial release we will take a conservative, high-level approach to overlap detection. Two backups will not run concurrently if there is any overlap in included namespaces. For example, if a backup that includes `ns1` and `ns2` is running, then a second backup for `ns2` and `ns3` will not be started. If a backup which does not filter namespaces is running (either a whole cluster backup or a non-namespace-limited backup with a label selector) then no other backups will be started, since a backup across all namespaces overlaps with any other backup. Calculating item-level overlap is problematic since we don't know which items are included in a backup until backup processing has begun.
+
+## Goals
+- Process multiple backups concurrently
+- Detect namespace overlap to avoid conflicts
+- For queued backups (not yet runnable due to concurrency limits or overlap), indicate the queue position in status
+
+## Non Goals
+- Item-level overlap detection
+- Handling NFS PVs when more than one PV point to the same underlying NFS share
+
+## High-Level Design
```

Comment:
Once we allow concurrent backups, below problems get more prominent:
1. The resource backup/restore may take more memory in a scaled env, so the current memory limit for Velero server is not enough
2. At present, Velero server enumerates the resources and stores them into the root fs (/tmp) of the pod, or into the root fs of the node. Concurrent backups may result in overrun of the node's root fs and finally causing Velero server pod to be evicted.

We need design for these two areas.

Reply:
@Lyndon-Li your concerns are valid. I agree with @sseago proposed solutions in both the cases. Documenting memory recommendations post-implementation sounds like the right approach.
For the /tmp concern, I like the idea of a configurable PVC mount, something like `--scratch-local-pvc myPVC` could give users flexibility in larger environments without changing default behavior.

Reply:
Previously I meant the problem of overrunning the root fs. If we don't want to solve it for the 1st delivery, let's document it.
A dedicated PVC could help on the problem, but things may be more complex:

- There would be a size of the volume
- We cannot predict the size of volume since this related to how large of the backups and how many concurrent backup
- So either in the root fs or in a dedicated volume, once this situation comes, all concurrent backups have to fail; or at least some of them must fail and release the space to solve this jam

Again, this is another case that the failure tolerance gets downgraded after concurrent backup involves.

Reply:
I think for the root fs, we can solve that here. The design calls for an optional installer param that takes the name of a PVC and, if present, Velero will mount /tmp in the velero pod using this PVC. Velero should not be involved in creating/managing the volume itself, since its configuration (size, storageclass, etc.) would be dependent on the user's environment, but if a user needs more space for /tmp, they can create and provision their own PVC in the velero namespace and supply this to the installer.

I don't really think we have to do anything about the concern around 2 backups overwriting each other's files, since we dynamically create new filenames via `os.CreateTemp` and store this path in the ItemCollector struct.

Reply:
But anything beyond taking the user-supplied PVC is out of scope for this release.

Reply:
I've pulled the user-supplied PVC out of the design. This will be resolved in the future by a more flexible option for Velero to create the PVC.